hal/src/rtl872x/exflash_hal.cpp

```diff
@@ -364,31 +372,288 @@ static bool isSecureOtpMode(uint32_t normalContent) {
+            memcpy(buf, &data, 4);
+            buf += 4;
+        }
+        cnt = size % 4;
+        if (cnt > 0) {
+            data = rtkSpiFlashGetDataReg(0, RTK_SPIC_DATA_WORD);
+            memcpy(buf, &data, cnt);
+            buf += cnt;
+        }
+    } else {
+        rtkSpiFlashEnableCsRead(RTK_SPIC_FIFO_HALF_SIZE);
+        cnt = RTK_SPIC_FIFO_HALF_SIZE / 4;
+        for (uint32_t i = 0; i < cnt; i++) {
+            data = rtkSpiFlashGetDataReg(0, RTK_SPIC_DATA_WORD);
+            memcpy(buf, &data, 4);
+            buf += 4;
+        }
+        /* t is transfer data times,
+         * t = (size - first xfer data) / fifo size
+         */
+        uint32_t t = (size - RTK_SPIC_FIFO_HALF_SIZE) / RTK_SPIC_FIFO_HALF_SIZE;
+        address = address + RTK_SPIC_FIFO_HALF_SIZE;
+        while (t > 0) {
+            rtkSpiFlashSendCmdAddr(cmd, address, dummyCycle);
+            rtkSpiFlashEnableCsRead(RTK_SPIC_FIFO_HALF_SIZE);
+            for (uint32_t i = 0; i < cnt; i++) {
+                data = rtkSpiFlashGetDataReg(0, RTK_SPIC_DATA_WORD);
+                memcpy(buf, &data, 4);
+                buf += 4;
+            }
+            address = address + RTK_SPIC_FIFO_HALF_SIZE;
+            t--;
+        }
+        uint32_t tmp = size % RTK_SPIC_FIFO_HALF_SIZE;
+        if (tmp > 0) {
+            rtkSpiFlashSendCmdAddr(cmd, address, dummyCycle);
+            rtkSpiFlashEnableCsRead(tmp);
+            cnt = tmp / 4;
+            for (uint32_t i = 0; i < cnt; i++) {
+                data = rtkSpiFlashGetDataReg(0, RTK_SPIC_DATA_WORD);
+                memcpy(buf, &data, 4);
+                buf += 4;
+            }
+            cnt = tmp % 4;
+            if (cnt > 0) {
+                data = rtkSpiFlashGetDataReg(0, RTK_SPIC_DATA_WORD);
+                memcpy(buf, &data, cnt);
+                buf += cnt;
+            }
+        }
+    }
+    dcacheCleanInvalidateAligned((uintptr_t)buf, size);
+    return SYSTEM_ERROR_NONE;
+}
+
+static int gd25WriteSecurityRegisters(uint8_t otpPageIdx, uint32_t addr, uint8_t* buf, size_t size) {
+    CHECK_TRUE(otpPageIdx < HAL_QSPI_FLASH_OTP_SECTOR_COUNT_GD25, SYSTEM_ERROR_INVALID_ARGUMENT);
+    CHECK_TRUE(addr + size <= HAL_QSPI_FLASH_OTP_SIZE_GD25, SYSTEM_ERROR_INVALID_ARGUMENT);
+
+    uint32_t address = ((otpPageIdx + 1) << 12) + addr;
+    uint8_t addrBuf[3] = {};
+    addrBuf[0] = (uint8_t)((address >> 16) & 0xFF);
+    addrBuf[1] = (uint8_t)((address >> 8) & 0xFF);
+    addrBuf[2] = (uint8_t)((address >> 0) & 0xFF);
+
+    // We may need to improve this
+    auto ptr = (uint8_t*)malloc(size + 3);
+    if (!ptr) {
+        return SYSTEM_ERROR_NO_MEMORY;
+    }
+    SCOPE_GUARD({
+        if (ptr) {
+            free(ptr);
+        }
+    });
+    memcpy(ptr, addrBuf, 3);
+    memcpy(ptr + 3, buf, size);
+
+    // Write/Erase enable
+    FLASH_TxCmd(GENERIC_FLASH_CMD_WREN, 0, nullptr);
+    FLASH_WaitBusy(WAIT_WRITE_EN);
+
+    FLASH_TxCmd(GD_FLASH_CMD_PGM_SEC, size + 3, ptr);
```

Comment:
⚠️ I seem to fault when writing >=10 bytes of data. 9 works fine, but anything more fails